src/ccache.c

```diff
@@ -0,0 +1,523 @@
+	++stats->nrequests;
+}
+
+/* Depends on booted tcache */
+bool
+ccache_init(tsdn_t *tsdn, base_t *base) {
+	assert(ncpus > 0);
+
+	ccache_minind = sz_size2index(tcache_maxclass) + 1;
+	ccache_maxind = ccache_minind + ccached_nclasses - 1;
+	ccache_maxclass = sz_index2size(ccache_maxind);
+	ccache_bin_size = PAGE;
+	ccache_size = sizeof(ccache_t) * ncpus;
+	ccache = (ccache_t*)base_alloc(tsdn, base, ccache_size,
+	    PAGE);
+	if (ccache == NULL) {
+		return true;
+	}
+	for (unsigned i = 0; i < ncpus; ++i) {
+		ccache_t *ith_cache = &ccache[i];
+		for (unsigned j = 0; j < ccached_nclasses; ++j) {
+			ccache_bin_t *bin = &ith_cache->bins[j];
+			bin->head = (void **)&bin->head;
+		}
+	}
+
+	return false;
+}
+
+void*
+ccache_alloc(tsd_t *tsd, arena_t *arena, size_t size, szind_t ind,
+    bool zero, bool small) {
+	/*
+	 * Ccache is disabled for manual arenas due to difficulties handling
+	 * arena reset.
+	 */
+	assert(arena_is_auto(arena));
+
+	void *ret = NULL;
+	volatile rseq_t *rseq_abi = &tsd_ccache_tdatap_get(tsd)->rseq_abi;
+	int cpu;
+	ccache_t *ccache;
+	ccache_bin_t *bin;
+restart:
+	cpu = rseq_abi->cpu_id;
+	ccache = ccache_get(cpu);
+	bin = ccache_bin_get(ccache, ind);
+	__asm__ __volatile__ goto(
+	    /* aw == Allocatable and writable section */
+	    ".pushsection __rseq_table, \"aw\"\n\t"
+	    ".balign 32\n\t"
+	    "cs_obj:\n\t"
+	    /* rseq_cs.version rseq_cs.flags */
+	    ".long 0, 0\n\t"
+	    /* rseq_cs.start_ip rseq_cs.post_commit_offset rseq_cs.abort_ip */
+	    ".quad rseq_cs_start%=, rseq_cs_postcommit%= - rseq_cs_start%=, rseq_cs_abort%=\n\t"
+	    ".popsection\n\t"
+
+	    "leaq cs_obj(%%rip), %%rax\n\t"
+	    "movq %%rax, %[rseq_cs]\n\t"
+
+	    /* -------------------- Begin ------------------- */
+	    "rseq_cs_start%=:\n\t"
+	    "cmpl %[cpu_id], %[initial_cpu_id]\n\t"
+	    /* Restart in case of CPU id missmatch*/
+	    "jne %l[restart]\n\t"
+
+	    /* rax = head */
+	    "movq %c[head_offset](%[bin]), %%rax\n\t"
+	    /* If in fallback mode, then jump to fallback label */
+	    "test %%rax, %%rax\n\t"
+	    "je %l[fallback]\n\t"
+	    /* If the bin is empty, then put into fallback mode and refill */
+	    "cmpq (%%rax), %%rax\n\t"
+	    "jne refill_not_needed%=\n\t"
+	    /* Prepare head = 0 */
+	    "xor %%rax, %%rax\n\t"
+	    "jmp commit%=\n\t"
+
+	    "refill_not_needed%=:\n\t"
+	    /* Store allocation in ret and preapare head++ */
+	    "movq (%%rax), %[ret]\n\t"
+	    "addq $8, %%rax\n\t"
+	    /* Commit head++ or head = 0 and exit CS */
+	    "commit%=:\n\t"
+	    "movq %%rax, %c[head_offset](%[bin])\n\t"
+
+	    /* ------------------- End ------------------- */
+	    "rseq_cs_postcommit%=:\n\t"
+
+	    /* ax == Allocatable and executable section */
+	    ".pushsection __rseq_abort, \"ax\"\n\t"
+	    ".byte 0x0f, 0x1f, 0x05\n\t"
+	    ".long "STRINGIFY(JEMALLOC_RSEQ_SIG)"\n\t"
+	    "rseq_cs_abort%=:\n\t"
+	    /*
+	     * Spin in case of abort - re-fetch per-cpu cache for the updated
```

Comment:
Curious about this: would it simplify things much, if on failure we simply give up / skip ccache?

Reply:
Don't think it would simplify things. The complexity stems from the concurrent access to the underlying data structure (stack), and the source of concurrent access is possible concurrent refilling and attempting to operate the stack. If we remove retries, that would mean if a thread is preempted during the critical section, it can't retry the access. This only removes one jump instruction to `restart` label, everything else stays the same.